File: java/src/io/zihengcat/leetcode/hmap/LongestPalindrome.java
package io.zihengcat.leetcode.hmap;

import java.util.Set;
import java.util.HashSet;

/**
 * LeetCode 409. Longest Palindrome
 * https://leetcode.com/problems/longest-palindrome/
 */
public class LongestPalindrome {
    public static void main(String[] args) {
        // ...
    }
    public int longestPalindrome(String s) {
        if (s == null) {
            return 0;
        }
        int cnt = 0;
        Set<Character> aSet = new HashSet<Character>();
        char c = 0;
        int n = s.length();
        for (int i = 0; i < n; i++) {
            c = s.charAt(i);
            if (aSet.contains(c)) {
                cnt++;
                aSet.remove(c);
            } else {
                aSet.add(c);
            }
        }
        return aSet.size() > 0
            ? 2 * cnt + 1
            : 2 * cnt;
    }
}

/* EOF */

File: cc/hash_table/LongestPalindrome.cc
#include <string>
#include <unordered_set>

using namespace std;

/**
 * LeetCode 409. Longest Palindrome
 * https://leetcode.com/problems/longest-palindrome/
 */
class LongestPalindrome {
public:
    int longestPalindrome(string s) {
        int cnt = 0;
        unordered_set<char> hset;
        unordered_set<char>::iterator iter;
        int n = s.size();
        for (int i = 0; i < n; i++) {
            iter = hset.find(s[i]);
            if (iter != hset.end()) {
                cnt++;
                hset.erase(iter);
            } else {
                hset.insert(s[i]);
            }
        }
        return hset.size() > 0
            ? 2 * cnt + 1
            : 2 * cnt;
    }
};

int main(int argc, char const *argv[]) {
    // ...
    return 0;
}

/* EOF */

